### PR TITLE
Submission of flexsdm and ENMTML

### PR DESCRIPTION
**Pull request by sjevelazco**

### CODE REVIEW

**Comment by sylvainschmitt** [comment]
Why did you closed your PR?

**Comment by sjevelazco** [comment]
I reopened it again.

**Comment by sylvainschmitt** [comment]
Thanks, to help review could you give an example of function in each package when you answered yes to the following fields:

- **occ_acquisition**: Function(s) to download occurrence data (yes/no).
- **occ_cleaning**: Function(s) to clean occurrence data (remove errors, fix georeferencing, etc.; yes/no)
- **data_integration**: Function(s) that statistically incorporate multiple types of data (e.g., presence/absence and presence-only; abundance and presence-only; etc.)
- **env_collinearity**: Function(s) to explore or address collinearity of environmental data (yes/no)
- **env_process**: Functions to process environmental data (interpolation, resampling, etc.; yes/no)
- **bias**: Function(s) to assess and/or correct biases in data (e.g., spatially thinning occurrences or weighting background to address occurrence sampling bias; yes/no)
- **study_region**: Function(s) to define the study area (yes/no)
- **backg_sample**: Function(s) to sample background/pseudoabsence records (yes/no)
- **data_partitioning**: Function(s) to partition data for model evaluation (yes/no)
- **mod_fit**: Function(s) to fit models (yes/no)
- **mod_ensemble**: Function(s) to generate ensemble models (yes/no)
- **mod_stack**: Function(s) to stack multiple single-species models to estimate community composition / diversity (yes/no)
- **mod_evaluate**: Function(s) to evaluate model performance (yes/no)
- **mod_multispecies**: Function(s) that model multiple species at once to estimate community composition / diversity (i.e., not a “stacked model; yes/no)
- **mod_mechanistic**: Function(s) that account for physiological, genetic, and/or trait-based components of species-environmental relationships (i.e., not just correlative; yes/no)
- **pred_general**: Function(s) to make model predictions (yes/no)
- **pred_extrapolation**: Function(s) to address or visualize prediction extrapolation (yes/no)
- **pred_inspect**: Function(s) to inspect or visualize the behavior of the model prediction (response curves, variable importance, xAI, etc.; yes/no)
- **post_processing**: Function(s) to post-process SDM predictions (dispersal or population simulation, niche overlap, estimation of biodiversity, etc.; yes/no)
- **gui**: Includes a graphical user interface (yes/no)
- **metadata**: Function(s) to  record analysis metadata (yes/no)

**Comment by sylvainschmitt** [comment]
@adamlilith or @jamiemkass could you review the metadata once @sjevelazco gave example functions? I only need one as I can also review for this first case.

**Comment by sylvainschmitt** [comment]
Have a look to #4 for an example of information needed. 
The automatic check validated the correspondence between your package metadata and the one given on CRAN.

**Comment by sjevelazco** [comment]
Okay, but in which field do I have to add that information?
None of the yaml files in https://github.com/sylvainschmitt/sdmverse/tree/main/inst/extdata/packages have examples of functions

**Comment by sylvainschmitt** [comment]
Sorry this was unclear. No just to help reviewers judge the fact you said yes to some categories, could you give them an example of at least one function or functionality of your packages that meet the fields you said your packages were dealing with. In your case it would be:

## ENMTML

- **occ_cleaning**: Function(s) to clean occurrence data (remove errors, fix georeferencing, etc.; yes/no)
- **env_collinearity**: Function(s) to explore or address collinearity of environmental data (yes/no)
- **bias**: Function(s) to assess and/or correct biases in data (e.g., spatially thinning occurrences or weighting background to address occurrence sampling bias; yes/no)
- **study_region**: Function(s) to define the study area (yes/no)
- **backg_sample**: Function(s) to sample background/pseudoabsence records (yes/no)
- **data_partitioning**: Function(s) to partition data for model evaluation (yes/no)
- **mod_fit**: Function(s) to fit models (yes/no)
- **mod_ensemble**: Function(s) to generate ensemble models (yes/no)
- **mod_evaluate**: Function(s) to evaluate model performance (yes/no)
- **mod_multispecies**: Function(s) that model multiple species at once to estimate community composition / diversity (i.e., not a “stacked model; yes/no)
- **pred_general**: Function(s) to make model predictions (yes/no)
- **pred_extrapolation**: Function(s) to address or visualize prediction extrapolation (yes/no)
- **post_processing**: Function(s) to post-process SDM predictions (dispersal or population simulation, niche overlap, estimation of biodiversity, etc.; yes/no)

## flexsdm

- **occ_cleaning**: Function(s) to clean occurrence data (remove errors, fix georeferencing, etc.; yes/no)
- **env_collinearity**: Function(s) to explore or address collinearity of environmental data (yes/no)
- **bias**: Function(s) to assess and/or correct biases in data (e.g., spatially thinning occurrences or weighting background to address occurrence sampling bias; yes/no)
- **study_region**: Function(s) to define the study area (yes/no)
- **backg_sample**: Function(s) to sample background/pseudoabsence records (yes/no)
- **data_partitioning**: Function(s) to partition data for model evaluation (yes/no)
- **mod_fit**: Function(s) to fit models (yes/no)
- **mod_ensemble**: Function(s) to generate ensemble models (yes/no)
- **mod_evaluate**: Function(s) to evaluate model performance (yes/no)
- **mod_multispecies**: Function(s) that model multiple species at once to estimate community composition / diversity (i.e., not a “stacked model; yes/no)
- **pred_general**: Function(s) to make model predictions (yes/no)
- **pred_extrapolation**: Function(s) to address or visualize prediction extrapolation (yes/no)
- **post_processing**: Function(s) to post-process SDM predictions (dispersal or population simulation, niche overlap, estimation of biodiversity, etc.; yes/no)

**Comment by sylvainschmitt** [comment]
Examples are to be given in this thread.

**Comment by sylvainschmitt** [comment]
@sjevelazco , could you please add examples as required for review? Otherwise your submission is ready to be accepted.

File: inst/extdata/packages/ENMTML.yaml
Comment (hannahlowens) [inline inst/extdata/packages/ENMTML.yaml]:
```suggestion
title: "Create and process ecological niche models, including several pre- and post-processing methods"
```

File: inst/extdata/packages/ENMTML.yaml
Comment (hannahlowens) [inline inst/extdata/packages/ENMTML.yaml:9]:
```suggestion
      process ecological niche models. The main
```

File: inst/extdata/packages/flexsdm.yaml
Comment (hannahlowens) [inline inst/extdata/packages/flexsdm.yaml]:
```suggestion
      post-processing."
```

**Review by hannahlowens — COMMENTED** [review]
Comments supplied are just suggestions for the package, not prohibitive RE: acceptance. Waiting for reply to [this request](https://github.com/sylvainschmitt/sdmverse/pull/3#issuecomment-1456155416) before approval.

**Comment by sylvainschmitt** [comment]
> Comments supplied are just suggestions for the package, not prohibitive RE: acceptance. Waiting for reply to [this request](https://github.com/sylvainschmitt/sdmverse/pull/3#issuecomment-1456155416) before approval.

As the package is not on CRAN you can indeed change the content of the fields accordingly if you want to @sjevelazco .

@hannahlowens , would you be available for a full review?

**Comment by hannahlowens** [comment]
Yep, can do!

**Comment by sjevelazco** [comment]
Hi Sylvain, Sorry for my delayed reply. Here are the examples for each category 

ENMTML
-----
**occ_cleaning**: ENMTML function will automatically remove any duplicated occurrence or georeferenced in empty cells 
**env_collinearity**:  the argument *colin_var* of ENMTML function performs different methods to reduce variable collinearity
**bias**:  the argument *thin_occ* of ENMTML function allows the use spatial filtering
**study_region**: different methods in the argument *sp_accessible_area* of ENMTML function
**backg_sample**: different methods in the argument *pseudoabs_method* of ENMTML function
**data_partitioning**: different methods in the argument *part* of ENMTML function
**mod_fit**: different methods in the argument *algorithm* of ENMTML function
**mod_ensemble**: different methods in the argument *ensemble* of ENMTML function
**mod_evaluate**: different methods for data partitioning *part* o use of external data *eval_occ* of ENMTML function
**mod_multispecies**: ENMTML function model any quantity of species at once 
**pred_general**:  ENMTML function could predict or not the model see *save_final* argument
**pred_extrapolation**: ENMTML function could evaluate extrapolation using *extrapolation* argument
**post_processing**: the argument MSDM if ENMTML function includes spatial restrictions to model projection. These methods restrict potential distributions to get models closer to occupied distributions.

flexsdm
---
**occ_cleaning**: function *env_outliers()*  detects outliers in the environmental space
**env_collinearity**: function *correct_colinvar()* performs collinearity reduction on predictors 
**bias**: *occfilt_env()* performs environmental filtering on species occurrences and *occfilt_geo()* performs geographical filtering on species occurrences
**study_region**: *calib_area()* delimits calibration area for constructing species distribution models **backg_sample**: *sample_background()* and *sample_pseudoabs()* samples background points or pseudo-absences 
**data_partitioning**: *part_random()* Data partitioning for training and testing models, *part_sblock()* Spatial block cross-validation, *part_sband()* Spatial band cross-validation, *part_senv()* Environmental cross-validation
**mod_fit**: *fit_*, *tuen_*, and *esm_* family functions
**mod_ensemble**: *fit_ensemble()* Fit and validate ensemble models with different ensemble methods
**mod_evaluate**: *sdm_eval()* Calculate different model performance metrics and *part_* functions perform different data partitioning methods
**pred_general**: *sdm_predict()* Spatial predictions of individual and ensemble mode
**pred_extrapolation**: *extra_eval()* Measure model extrapolation, *extra_truncate()* Constraint suitability values under a given extrapolation value 
**post_processing**: *msdm_posteriori()* Methods to correct overprediction of species distribution models based on occurrences and suitability patterns, *extra_truncate()* Constraint suitability values under a given extrapolation value

**Comment by sjevelazco** [comment]
Hi Sylvain, Sorry for my delayed reply. Here are the examples for each category 

ENMTML
-----
**occ_cleaning**: ENMTML function will automatically remove any duplicated occurrence or georeferenced in empty cells 
**env_collinearity**:  the argument *colin_var* of ENMTML function performs different methods to reduce variable collinearity
**bias**:  the argument *thin_occ* of ENMTML function allows the use spatial filtering
**study_region**: different methods in the argument *sp_accessible_area* of ENMTML function
**backg_sample**: different methods in the argument *pseudoabs_method* of ENMTML function
**data_partitioning**: different methods in the argument *part* of ENMTML function
**mod_fit**: different methods in the argument *algorithm* of ENMTML function
**mod_ensemble**: different methods in the argument *ensemble* of ENMTML function
**mod_evaluate**: different methods for data partitioning *part* o use of external data *eval_occ* of ENMTML function
**mod_multispecies**: ENMTML function model any quantity of species at once 
**pred_general**:  ENMTML function could predict or not the model see *save_final* argument
**pred_extrapolation**: ENMTML function could evaluate extrapolation using *extrapolation* argument
**post_processing**: the argument MSDM if ENMTML function includes spatial restrictions to model projection. These methods restrict potential distributions to get models closer to occupied distributions.

flexsdm
---
**occ_cleaning**: function *env_outliers()*  detects outliers in the environmental space
**env_collinearity**: function *correct_colinvar()* performs collinearity reduction on predictors 
**bias**: *occfilt_env()* performs environmental filtering on species occurrences and *occfilt_geo()* performs geographical filtering on species occurrences
**study_region**: *calib_area()* delimits calibration area for constructing species distribution models **backg_sample**: *sample_background()* and *sample_pseudoabs()* samples background points or pseudo-absences 
**data_partitioning**: *part_random()* Data partitioning for training and testing models, *part_sblock()* Spatial block cross-validation, *part_sband()* Spatial band cross-validation, *part_senv()* Environmental cross-validation
**mod_fit**: *fit_*, *tuen_*, and *esm_* family functions
**mod_ensemble**: *fit_ensemble()* Fit and validate ensemble models with different ensemble methods
**mod_evaluate**: *sdm_eval()* Calculate different model performance metrics and *part_* functions perform different data partitioning methods
**pred_general**: *sdm_predict()* Spatial predictions of individual and ensemble mode
**pred_extrapolation**: *extra_eval()* Measure model extrapolation, *extra_truncate()* Constraint suitability values under a given extrapolation value 
**post_processing**: *msdm_posteriori()* Methods to correct overprediction of species distribution models based on occurrences and suitability patterns, *extra_truncate()* Constraint suitability values under a given extrapolation value

**Comment by sylvainschmitt** [comment]
Thanks @sjevelazco . @hannahlowens we can now proceed to review with the given information.

File: inst/extdata/packages/ENMTML.yaml
Comment (sylvainschmitt) [inline inst/extdata/packages/ENMTML.yaml]: >  mod_multispecies: ENMTML function model any quantity of species at once

Yes but this only parrallel computing of SDMs, but is ENMTML providing any multispecies outputs?  See documentation:

> mod_multispecies: Function(s) that model multiple species at once to estimate community composition / diversity (i.e., not a “stacked model; yes/no)

If not please change by "no" because we try to discriminate packages metadata and we will accept fields with a yes if it is really a major functionality of the package.

File: inst/extdata/packages/ENMTML.yaml
Comment (sylvainschmitt) [inline inst/extdata/packages/ENMTML.yaml:32]: >  mod_evaluate: different methods for data partitioning part o use of external data eval_occ of ENMTML function

This is unclear to me, is this partitioning or evaluation?

**Review by sylvainschmitt — CHANGES_REQUESTED** [review]
@sjevelazco , thanks again for your submission. Please answer to my comment below for ENMTML on mod_evaluate and mod_multispecies before acceptance.

File: inst/extdata/packages/ENMTML.yaml
Comment (hannahlowens) [inline inst/extdata/packages/ENMTML.yaml]:
```suggestion
bias: no
```
Argument `thin_occ` is within function `ENMTML()`. I'd characterize thinning as "cleaning" more than assessing bias. @sylvainschmitt, what do you think?

File: inst/extdata/packages/flexsdm.yaml
Comment (hannahlowens) [inline inst/extdata/packages/flexsdm.yaml]:
```suggestion
bias: no
```
Again, these functions seem more like occurrence cleaning than bias, but I'm not clear on the line @sylvainschmitt has conceptualized.

File: inst/extdata/packages/ENMTML.yaml
Comment (sylvainschmitt) [reply inst/extdata/packages/ENMTML.yaml]: Yes I was hesitating on this one too. And I think you are right. In my case for `SSDM` we rely on `spThin` for spatial thinning but I did not include it in the bias field: https://github.com/sylvainschmitt/sdmverse/blob/cf4e0f8e23db8b1ee024d46cf240daa196a371f2/inst/extdata/packages/SSDM.yaml#L22
Indeed thinning is more cleaning occurrences than assessing bias as you pointed out. Secondly, the aim of sdmverse is to illustrate the complementarity and differences among packages. So I think we should focus on core functionalities, and avoid answer yes for fields when it is only minor functionalities (because if everybody have yes everywhere sdmverse becomes useless).

Anyway this is great that you pointed it out, and it illustrate how it is important that we always try to have at least two reviewers.

**Review by hannahlowens — CHANGES_REQUESTED** [review]
Please remove "bias" from both packages.

File: inst/extdata/packages/ENMTML.yaml
Comment (hannahlowens) [reply inst/extdata/packages/ENMTML.yaml]: I'm liking this system. Thanks for setting it up.

File: inst/extdata/packages/ENMTML.yaml
Comment (sjevelazco) [reply inst/extdata/packages/ENMTML.yaml]: [sylvainschmitt](https://github.com/sylvainschmitt) *Yes but this only parrallel computing of SDMs, but is ENMTML providing any multispecies outputs?*
Okay, so mod_multispecies =  No

File: inst/extdata/packages/ENMTML.yaml
Comment (sjevelazco) [reply inst/extdata/packages/ENMTML.yaml:32]: Both different partition methods and different ways to validate models

**Comment by jamiemkass** [comment]
Sorry all -- just started a new job and have been incredibly busy. Will make sure to review more of these from next week on.

**Comment by sylvainschmitt** [comment]
> Sorry all -- just started a new job and have been incredibly busy. Will make sure to review more of these from next week on.

no worry, @hannahlowens has been super-active.

**Comment by jamiemkass** [comment]
Thanks for picking up my slack, @hannahlowens! I owe you~

**Comment by hannahlowens** [comment]
> Thanks for picking up my slack, @hannahlowens! I owe you~

**Comment by sylvainschmitt** [comment]
@sjevelazco I fixed your submission following reviews. After merging your pull request, you will be added to the repository contributors. Please expect to be contacted for a review request in the future.

**Comment by sjevelazco** [comment]
Hi @sylvainschmitt 
I noticed that flexsdm has no model tune feature in sdmverse. However, flexsdm has a family function called [tune_](https://sjevelazco.github.io/flexsdm/#tune_-functions-family) for modeling tuning. Please, can you add this flexsdm feature in sdmverse?

Thanks

**Pull request merged**
